[BUG] cudf.testing methods do not respect tolerances

**Describe the bug**
cudf's testing utilities only find two frame-like objects to be equal if they are exactly equal, despite accepting `atol` and `rtol` parameters. The problem is that these parameters are ignored by `cudf.testing.assert_column_equal`, which simply calls `lhs.equal(rhs)`.

**Steps/Code to reproduce bug**
```
>>> import cudf
>>> import pandas as pd
>>> eps = 1e-4
>>> cudf.testing.assert_series_equal(cudf.Series([1e6 + eps]), cudf.Series([1e6 + 2*eps]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nfs/vyasr/local/rapids/cudf/python/cudf/cudf/testing/testing.py", line 489, in assert_series_equal
    atol=atol,
  File "/home/nfs/vyasr/local/rapids/cudf/python/cudf/cudf/testing/testing.py", line 222, in assert_column_equal
    obj, f"values are different ({np.round(diff, 5)} %)", msg1, msg2,
  File "/home/nfs/vyasr/local/rapids/cudf/python/cudf/cudf/testing/testing.py", line 37, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: ColumnBase are different

values are different (100.0 %)
[left]:  [1000000.0001]
[right]: [1000000.0002]
>>> pd.testing.assert_series_equal(pd.Series([1e6 + eps]), pd.Series([1e6 + 2*eps]))
>>>
```

**Expected behavior**
The first assertion should succeed for a sufficiently small value of eps (at least, larger than machine epsilon, depending on what the smallest representable number is in cudf/pandas). 

**Additional context**
Since #7495 removed the `assert_eq` function from `cudf`'s public namespace, fixing this problem is a blocker for downstream packages that need to adapt tests to make use of these functions.

--- a/python/cudf/cudf/testing/testing.py
+++ b/python/cudf/cudf/testing/testing.py
@@ -4,6 +4,7 @@ from __future__ import annotations
 
 from typing import Union
 
+import cupy as cp
 import numpy as np
 import pandas as pd
 
@@ -203,7 +204,11 @@ def assert_column_equal(
 
     columns_equal = False
     try:
-        columns_equal = left.equals(right)
+        columns_equal = (
+            left.equals(right)
+            if check_exact
+            else cp.isclose(left.values, right.values, rtol=rtol, atol=atol)
+        )
     except TypeError as e:
         if str(e) != "Categoricals can only compare with the same type":
             raise e

--- a/python/cudf/cudf/testing/testing.py
+++ b/python/cudf/cudf/testing/testing.py
@@ -9,6 +9,7 @@ import numpy as np
 import pandas as pd
 
 import cudf
+from cudf.api.types import is_numeric_dtype
 from cudf.core._compat import PANDAS_GE_110
 from cudf.utils.dtypes import is_categorical_dtype
 
@@ -205,9 +206,15 @@ def assert_column_equal(
     columns_equal = False
     try:
         columns_equal = (
-            left.equals(right)
-            if check_exact
-            else cp.allclose(left.values, right.values, rtol=rtol, atol=atol)
+            (
+                cp.all(left.isnull().values == right.isnull().values)
+                and cp.allclose(
+                    left[left.isnull().unary_operator("not")].values,
+                    right[right.isnull().unary_operator("not")].values,
+                )
+            )
+            if not check_exact and is_numeric_dtype(left)
+            else left.equals(right)
         )
     except TypeError as e:
         if str(e) != "Categoricals can only compare with the same type":

--- a/python/cudf/cudf/testing/testing.py
+++ b/python/cudf/cudf/testing/testing.py
@@ -207,7 +207,7 @@ def assert_column_equal(
         columns_equal = (
             left.equals(right)
             if check_exact
-            else cp.isclose(left.values, right.values, rtol=rtol, atol=atol)
+            else cp.allclose(left.values, right.values, rtol=rtol, atol=atol)
         )
     except TypeError as e:
         if str(e) != "Categoricals can only compare with the same type":

